app/src/apiService.ts

```diff
@@ -0,0 +1,53 @@
+import axios, { AxiosError, AxiosInstance } from "axios";
+import {
+    createAsyncThunk,
+    AsyncThunkOptions,
+    AsyncThunk,
+} from "@reduxjs/toolkit";
+
+const baseURL = "http://localhost:8080";
+
+interface CustomAsyncThunkOptions<E>
+    extends AsyncThunkOptions<void, { rejectValue: E }> {
+    rejectValue: E;
+}
+
+interface API<T, E> {
+    get(
+        endpoint: string,
+        asyncThunkConfig?: CustomAsyncThunkOptions<E>
+    ): AsyncThunk<T, void, CustomAsyncThunkOptions<E>>;
+}
+
+export class CreateAsyncThunk<T, E> implements API<T, E> {
+    private readonly type: string;
+    private readonly axiosInstance: AxiosInstance;
+
+    constructor(type: string) {
+        this.type = type;
+        this.axiosInstance = axios.create({ baseURL });
+    }
+
+    get = (
+        endpoint: string,
+        asyncThunkConfig?: CustomAsyncThunkOptions<E>
+    ): AsyncThunk<T, void, CustomAsyncThunkOptions<E>> =>
+        createAsyncThunk<T, void, CustomAsyncThunkOptions<E>>(
+            this.type,
+            async (_, thunkAPI) => {
+                try {
+                    const response = await this.axiosInstance.get<T>(endpoint);
+                    return response.data;
+                } catch (error) {
+                    let errorMessage = null;
+                    if (error instanceof AxiosError && error.response) {
+                        errorMessage = error.response.data;
+                    }
+                    return thunkAPI.rejectWithValue(errorMessage);
+                }
+            },
+            asyncThunkConfig
+        );
+}
+
+export const api = <T, E>(type: string) => new CreateAsyncThunk<T, E>(type);
```

Comment:
This doesn't feel quite right; this class isn't a general API client but a way of generating an action with given type. It'd be good to end up with an interface that looks more like, in ./packets/thunks:

```
import { createAsyncThunk } from '@reduxjs/toolkit'
import { api } from './api'

const FETCH = 'packets/fetch'
const fetchpackets = createAsyncThunk(
  FETCH,
  async (thunkAPI) => {
    const response = await api.get("/packets")
    return response.data
  }
)
```

where `api.get` would be a method that returns a promise, something like
```

 get(endpoint: string): Promise<T> {
       return this.axiosInstance.get<T>(endpoint)
                    .then(response => {
                        return response.data;
                    }).catch(error => {
                    let errorMessage = "Could not parse API response";
                    if (error instanceof AxiosError && error.response) {
                        errorMessage = error.response.data;
                    }
                    throw Error(errorMessage);
                })
 }
```

I think if I understand the redux docs correctly, no need to call `thunkAPI.rejectWithValue` directly, it will just reject with the thrown error message.